### ISSUE TITLE
Implement testing infrastructure for Gutenberg integrations

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#4108]
Based on #4107, this issue is focused on setting up the necessary infrastructure to test Gutenberg integrations as reliably as possible.

Some particular challenges to work around here include:
* Gutenberg has been included in core since WordPress 5.0. Technically even older versions could use the Gutenberg plugin though, which makes things more complicated. Or even any current WordPress version can use the Gutenberg plugin in an even newer version to get more of the "bleeding-edge" changes.
* Gutenberg APIs have been modified over time and unfortunately are by far not as stable/backward-compatible as WordPress core's PHP APIs.

We need a testing infrastructure that caters for this. While it is impossible to cover all possible combinations of versions, we should have a good set to rely on. Potentially, we should introduce a subset of our e2e test suite specifically for Gutenberg-based e2e tests, where we would run for a few more scenarios (e.g. one pre-5.0 with Gutenberg plugin, one with 5.0, one with 5.3, one with latest). In addition to e2e tests (which will be the most reliable automatic way to test the integrations), a good additional measure could be to always use the exact versions used in WordPress 5.0 for all the Gutenberg dependencies that Site Kit is using even in Jest tests (at least this part will technically rely on #4107 to be completed first) - to ensure we don't use APIs/functions that were only supported at a later point.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

### Test Coverage

* 

## QA Brief

* 

## Changelog entry

*

**Comment by felixarntz** [comment]
Thoughts on this one @aaemnnosttv @tofumatt @eugene-manuilov? Let's prioritize this for the new year to find a good approach.

**Comment by felixarntz** [comment]
@aaemnnosttv @tofumatt @eugene-manuilov Updated ACs for your review. If y'all agree, let's move it to IB.

**Comment by aaemnnosttv** [comment]
LGTM 👍

**Comment by FlicHollis** [comment]
Hi @tofumatt notice this has been picked up a little while ago. Are you still planning on working on this IB? If not might be good to un-assign yourself and let someone else pick up. Let me know what you think, Thanks!

**Comment by tofumatt** [comment]
I've updated the ACs here to omit the section around WordPress versions before 5.2, since 5.2 is now our minimum supported WordPress version. (See: #5874)

**Comment by aaemnnosttv** [comment]
@felixarntz does this make sense to prioritize right now seeing as we no longer have a planned integration in the block editor?

**Comment by felixarntz** [comment]
@aaemnnosttv Good question, it's still a good thing to work on, but not a P0 anymore. I'll update to a P1 ("Infrastructure" already has some notion of that this is not as important as e.g. a P1 bug or enhancement).

**Comment by techanvil** [comment]
Thanks @tofumatt, this IB is looking good. I have a few questions:

- When you mention our existing "enable plugin" approach, do you mean using the `activatePlugin` function from `@wordpress/e2e-test-utils`? If so, it could be worth linking an example of our use of it, to help illustrate the point. If not, it could use a bit of clarification.
- You mentioned exploring options during the Hackathon, would this exploration be worth linking to, if possible?
- I am not sure I fully understand the rationale behind converting our existing Idea Hub Notice integration to a React component. I can see the benefit of migrating to make more idiomatic use of APIs, but is it a necessity for the task in hand? If not, maybe it could it be split out into a separate issue, seeing as this one is a bit on the hefty side.

**Comment by tofumatt** [comment]
Indeed, I meant using `activatePlugin` to test with and without the latest Gutenberg plugin enabled. If possible we'd want to download the latest version of the plugin, or setup a system where we routinely include the latest version in our E2E plugins if that's not possible. I added a note about the helper usage though with an example 👍🏻 

In terms of options to link to—nothing really as most of my exploration with anything more complex/complicated (eg. a dedicated set of Gutenberg E2E separate from the rest of the tests with a dedicated runner, etc.) all proven complicated/cumbersome to run and unsatisfactory. The summary I think I posted to the Hackathon slides was basically that our original ideas of creating a "dedicated" set of E2E tests for Gutenberg environments would be at best ergonomically awkward and more work to maintain, but I couldn't even get it working well compared to just adding tests that would run as part of the existing E2E suite which was straightforward. Given our minimum WP version always supports Gutenberg that's now easy to do so it's what I suggest. Sorry there's not much to show, but it'd mostly just be half-baked failing branches that never got off the ground 😅 

Converting the existing integration is largely around being able to use more idiomatic APIs. I actually don't think that's a huge change given the APIs we'd use would be `useSelect` so straightforward to add… and the only other thing we'd need to do is a `React.render()` call for the div we use for that Idea Hub integration. It's not strictly required as part of this issue, fair enough, and I suppose we could add the test for that "component" _without_ migrating it and then migrate it later—even benefitting from the E2E test being there if you like. I figured it was a relatively small part of this issue, but if you still think it's best to separate out let me know and I'll file a follow-up issue for that.

**Comment by techanvil** [comment]
Hey @tofumatt, thanks for clarifying, and making that addition to the IB.

With regard to the Idea Hub notice, I have had a more fundamental realisation - we are in fact removing Idea Hub, via https://github.com/google/site-kit-wp/issues/6235, so for one thing it's clearly not worth the effort of converting the integration when its about to be removed.

It does also bring into question the viability of this Gutenberg testing framework if we don't have any tests for it, unless there are some on the horizon. Do you know if there's anything planned that would make use of it? I am a bit wary about adding code that's not being used at all...

**Comment by FlicHollis** [comment]
Thank you @techanvil and @tofumatt I will remove from the sprint too.
